File: src/common/test-authorize.c
/*
 * SPDX-License-Identifier: BSD-3-Clause
 * Copyright (c) 2014 Red Hat Inc.
 * Author: Stef Walter <stefw@redhat.com>
 */

#include "config.h"

#include "cockpitauthorize.h"
#include "testlib/cockpittest.h"

#include <sys/types.h>
#include <sys/wait.h>

#include <err.h>
#include <errno.h>
#include <pwd.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

static const char *expect_message;
static char *user;

static void
test_logger (const char *msg)
{
  g_assert_nonnull (msg);

  if (expect_message)
    {
      g_assert_nonnull (strstr (msg, expect_message));
      expect_message = NULL;
    }
  else
    {
      warnx ("%s", msg);
    }
}

static void
setup (char **user_ptr,
       gconstpointer data)
{
  struct passwd *pw;

  expect_message = NULL;

  pw = getpwuid (getuid ());
  g_assert_nonnull (pw);
  *user_ptr = g_strdup (pw->pw_name);
  user = *user_ptr;
}

static void
teardown (char **user_ptr,
          gconstpointer data)
{
  if (expect_message)
    g_assert_not_reached ();
  g_free (*user_ptr);
  user = NULL;
}

typedef struct {
  const char *input;
  const char *expected;
  const char *ret;
  int errn;
} ChallengeFixture;

static ChallengeFixture type_fixtures[] = {
  { "valid", "valid", "", 0 },
  { ":invalid", NULL, NULL, EINVAL },
  { "Basic more-data", "basic", "more-data", 0 },
  { "Basic   more-data", "basic", "more-data", 0 },
  { "valid:test", "valid", "test", 0 },
  { "valid1:", "valid1", "", 0 },
  { "valid2:test:test", "valid2", "test:test", 0 },
  { NULL },
};

static void
test_type (char **user_ptr,
           gconstpointer data)
{
  const ChallengeFixture *fix = data;
  const char *result;
  g_autofree char *type = NULL;

  result = cockpit_authorize_type (fix->input, &type);
  if (fix->errn != 0)
    g_assert_cmpint (errno, ==, fix->errn);
  if (fix->ret)
    {
      g_assert_cmpstr (result, ==, fix->ret);
      g_assert_cmpstr (type, ==, fix->expected);
    }
  else
    {
      g_assert_null (result);
    }
}

static ChallengeFixture subject_fixtures[] = {
  { "valid:73637275666679:", "73637275666679", "", 0 },
  { "valid:73637275666679:more-data", "73637275666679", "more-data", 0 },
  { "valid:scruffy:", "scruffy", "", 0 },
  { "X-Conversation conversationtoken more-data", "conversationtoken", "more-data", 0 },
  { "X-Conversation  conversationtoken    more-data", "conversationtoken", "more-data", 0 },
  { "invalid:", "73637275666679", NULL, EINVAL },
  { "invalid", NULL, NULL, EINVAL },
  { NULL },
};

static void
test_subject (char **user_ptr,
              gconstpointer data)
{
  const ChallengeFixture *fix = data;
  const char *result;
  g_autofree char *subject = NULL;

  if (fix->ret == NULL)
    expect_message = "\"authorize\" message";

  result = cockpit_authorize_subject (fix->input, &subject);
  if (fix->errn != 0)
    g_assert_cmpint (errno, ==, fix->errn);
  if (fix->ret)
    {
      g_assert_cmpstr (result, ==, fix->ret);
      g_assert_cmpstr (subject, ==, fix->expected);
    }
  else
    {
      g_assert_null (result);
    }
}

static ChallengeFixture basic_fixtures[] = {
  { "Basic c2NydWZmeTp6ZXJvZw==", "scruffy", "zerog", 0 },
  { "Basic!c2NydWZmeTp6ZXJvZw==", NULL, NULL, EINVAL },
  { "Basic c2NydWZ!!eXplcm9n", NULL, NULL, EINVAL },
  { "Basic c2NydWZmeXplcm9n", NULL, NULL, EINVAL },
  { "Basic", NULL, "", 0 },
  { NULL },
};

static void
test_parse_basic (char **user_ptr,
                  gconstpointer data)
{
  const ChallengeFixture *fix = data;
  const char *original = "blah";
  char *user = (char *)original;
  g_autofree char *password = NULL;

  if (fix->ret == NULL)
    expect_message = "invalid";

  password = cockpit_authorize_parse_basic (fix->input, &user);
  if (fix->errn != 0)
    g_assert_cmpint (errno, ==, fix->errn);
  if (fix->ret)
    {
      g_assert_cmpstr (password, ==, fix->ret);
      if (fix->expected)
        g_assert_cmpstr (user, ==, fix->expected);
      else
        g_assert_null (user);
    }
  else
    {
      g_assert_null (password);
      g_assert_cmpstr (user, ==, "blah"); /* not reassigned */
    }

  if (user != original)
    g_free (user);
  g_clear_pointer (&password, g_free);

  if (fix->ret == NULL)
    expect_message = "invalid";

  password = cockpit_authorize_parse_basic (fix->input, NULL);
  if (fix->errn != 0)
    g_assert_cmpint (errno, ==, fix->errn);
  if (fix->ret)
    {
      g_assert_cmpstr (password, ==, fix->ret);
    }
  else
    {
      g_assert_null (password);
    }
}

typedef struct {
  const char *input;
  size_t length;
  const char *ret;
  int errn;
} NegotiateFixture;

static NegotiateFixture parse_negotiate_fixtures[] = {
  { "Negotiate c2NydWZmeTp6ZXJvZw==", 13, "scruffy:zerog", 0 },
  { "Negotiate!c2NydWZmeTp6ZXJvZw==", 0, NULL, EINVAL },
  { "Negotiate c2Nyd!!ZmeTp6ZXJvZw==", 0, NULL, EINVAL },
  { "Negotiate", 0, "", 0 },
  { NULL },
};

static void
test_parse_negotiate (char **user_ptr,
                      gconstpointer data)
{
  const NegotiateFixture *fix = data;
  size_t length = 0xFFFFDD;
  g_autofree void *result = NULL;

  if (fix->ret == NULL)
    expect_message = "invalid";

  result = cockpit_authorize_parse_negotiate (fix->input, &length);
  if (fix->errn != 0)
    g_assert_cmpint (errno, ==, fix->errn);
  if (fix->ret)
    {
      g_assert_cmpuint (fix->length, ==, length);
      g_assert_cmpint (memcmp (result, fix->ret, length), ==, 0);
    }
  else
    {
      g_assert_null (result);
      g_assert_cmpuint (length, ==, 0xFFFFDD); /* not reassigned */
    }

  g_clear_pointer (&result, g_free);

  if (fix->ret == NULL)
    expect_message = "invalid";

  result = cockpit_authorize_parse_negotiate (fix->input, NULL);
  if (fix->errn != 0)
    g_assert_cmpint (errno, ==, fix->errn);
  if (fix->ret)
    {
      g_assert_cmpint (memcmp (result, fix->ret, fix->length), ==, 0);
    }
  else
    {
      g_assert_null (result);
    }
}

static NegotiateFixture build_negotiate_fixtures[] = {
  { "scruffy:zerog", 13, "Negotiate c2NydWZmeTp6ZXJvZw==", 0 },
  { NULL, 0, "Negotiate", 0, },
  { NULL },
};

static void
test_build_negotiate (char **user_ptr,
                      gconstpointer data)
{
  const NegotiateFixture *fix = data;
  g_autofree char *result = NULL;

  if (fix->ret == NULL)
    expect_message = "invalid";

  result = cockpit_authorize_build_negotiate (fix->input, fix->length);
  if (fix->errn != 0)
    g_assert_cmpint (errno, ==, fix->errn);
  if (fix->ret)
    {
      g_assert_cmpstr (result, ==, fix->ret);
    }
  else
    {
      g_assert_null (result);
    }
}

typedef struct {
  const char *input;
  const char *conversation;
  const char *ret;
  int errn;
} XConversationFixture;

static XConversationFixture parse_x_conversation_fixtures[] = {
  { "X-Conversation abcdefghi c2NydWZmeTp6ZXJvZw==", NULL, "scruffy:zerog", 0 },
  { "X-Conversation abcdefghi", NULL, "", 0 },
  { "X-Conversation abcdefghi c2NydW!!meTp6ZXJvZw==", NULL, NULL, EINVAL },
  { NULL },
};

static void
test_parse_x_conversation (char **user_ptr,
                           gconstpointer data)
{
  const XConversationFixture *fix = data;
  g_autofree char *result = NULL;

  if (fix->ret == NULL)
    expect_message = "invalid";

  result = cockpit_authorize_parse_x_conversation (fix->input, NULL);
  if (fix->errn != 0)
    g_assert_cmpint (errno, ==, fix->errn);
  if (fix->ret)
    {
      g_assert_cmpstr (result, ==, fix->ret);
    }
  else
    {
      g_assert_null (result);
    }
}

static XConversationFixture build_x_conversation_fixtures[] = {
  { "scruffy:zerog", "abcdefghi", "X-Conversation abcdefghi c2NydWZmeTp6ZXJvZw==", 0 },
  { "scruffy:zerog", NULL, " c2NydWZmeTp6ZXJvZw==", 0 },
  { "", "abcdefghi", "X-Conversation abcdefghi", 0 },
  { "scruffy:zerog", "", NULL, EINVAL },
  { NULL },
};

static void
test_build_x_conversation (char **user_ptr,
                           gconstpointer data)
{
  const XConversationFixture *fix = data;
  g_autofree char *conversation = NULL;
  g_autofree char *result = NULL;

  if (fix->ret == NULL)
    expect_message = "invalid";

  if (fix->conversation)
    conversation = g_strdup (fix->conversation);

  result = cockpit_authorize_build_x_conversation (fix->input, &conversation);
  if (fix->errn != 0)
    g_assert_cmpint (errno, ==, fix->errn);
  if (fix->ret)
    {
      if (strstr (fix->ret, "X-Conversation"))
        g_assert_cmpstr (result, ==, fix->ret);
      else
        g_assert_nonnull (strstr (result, fix->ret));
    }
  else
    {
      g_assert_null (result);
    }
}

int
main (int argc,
      char *argv[])
{
  int i;

  /* Some initial preparation */
  signal (SIGPIPE, SIG_IGN);
  cockpit_authorize_logger (test_logger, 0);

  cockpit_test_init (&argc, &argv);

  for (i = 0; type_fixtures[i].input != NULL; i++)
    {
      g_autofree gchar *name = g_strdup_printf ("/authorize/type/%s", type_fixtures[i].input);
      g_test_add (name, char *, type_fixtures + i,
                  setup, test_type, teardown);
    }
  for (i = 0; subject_fixtures[i].input != NULL; i++)
    {
      g_autofree gchar *name = g_strdup_printf ("/authorize/subject/%s", subject_fixtures[i].input);
      g_test_add (name, char *, subject_fixtures + i,
                  setup, test_subject, teardown);
    }
  for (i = 0; basic_fixtures[i].input != NULL; i++)
    {
      g_autofree gchar *name = g_strdup_printf ("/authorize/basic/%s", basic_fixtures[i].input);
      g_test_add (name, char *, basic_fixtures + i,
                  setup, test_parse_basic, teardown);
    }
  for (i = 0; parse_negotiate_fixtures[i].input != NULL; i++)
    {
      g_autofree gchar *name = g_strdup_printf ("/authorize/negotiate/parse/%s", parse_negotiate_fixtures[i].input);
      g_test_add (name, char *, parse_negotiate_fixtures + i,
                  setup, test_parse_negotiate, teardown);
    }
  for (i = 0; build_negotiate_fixtures[i].ret != NULL; i++)
    {
      g_autofree gchar *name = g_strdup_printf ("/authorize/negotiate/build/%s", build_negotiate_fixtures[i].ret);
      g_test_add (name, char *, build_negotiate_fixtures + i,
                  setup, test_build_negotiate, teardown);
    }
  for (i = 0; parse_x_conversation_fixtures[i].input != NULL; i++)
    {
      g_autofree gchar *name = g_strdup_printf ("/authorize/x-conversation/parse/%s", parse_x_conversation_fixtures[i].input);
      g_test_add (name, char *, parse_x_conversation_fixtures + i,
                  setup, test_parse_x_conversation, teardown);
    }
  for (i = 0; build_x_conversation_fixtures[i].input || build_x_conversation_fixtures[i].ret; i++)
    {
      g_autofree gchar *name = g_strdup_printf ("/authorize/x-conversation/build/%d", i);
      g_test_add (name, char *, build_x_conversation_fixtures + i,
                  setup, test_build_x_conversation, teardown);
    }

  return g_test_run ();
}

File: test/pytest/ws/test_authorize.py
# Copyright (C) 2014 Red Hat, Inc.
# SPDX-License-Identifier: GPL-3.0-or-later
# Author: Stef Walter <stefw@redhat.com>

import pytest

from test.pytest.ws import authorize


@pytest.mark.parametrize(
    ("challenge", "expected_type", "expected_remainder"),
    [
        ("valid", "valid", ""),
        ("Basic more-data", "basic", "more-data"),
        ("Basic   more-data", "basic", "more-data"),
        ("valid:test", "valid", "test"),
        ("valid1:", "valid1", ""),
        ("valid2:test:test", "valid2", "test:test"),
    ],
)
def test_parse_type(
    challenge: str, expected_type: str, expected_remainder: str
) -> None:
    auth_type, remainder = authorize.parse_type(challenge)
    assert auth_type == expected_type
    assert remainder == expected_remainder


@pytest.mark.parametrize(
    "challenge",
    [
        ":invalid",
    ],
)
def test_parse_type_invalid(challenge: str) -> None:
    with pytest.raises(ValueError):
        authorize.parse_type(challenge)


@pytest.mark.parametrize(
    ("challenge", "expected_subject", "expected_remainder"),
    [
        ("valid:73637275666679:", "73637275666679", ""),
        ("valid:73637275666679:more-data", "73637275666679", "more-data"),
        ("valid:scruffy:", "scruffy", ""),
        (
            "X-Conversation conversationtoken more-data",
            "conversationtoken",
            "more-data",
        ),
        (
            "X-Conversation  conversationtoken    more-data",
            "conversationtoken",
            "more-data",
        ),
    ],
)
def test_parse_subject(
    challenge: str, expected_subject: str, expected_remainder: str
) -> None:
    subject, remainder = authorize.parse_subject(challenge)
    assert subject == expected_subject
    assert remainder == expected_remainder


@pytest.mark.parametrize(
    "challenge",
    [
        "invalid:",
        "invalid",
    ],
)
def test_parse_subject_invalid(challenge: str) -> None:
    with pytest.raises(ValueError):
        authorize.parse_subject(challenge)


@pytest.mark.parametrize(
    ("response", "expected_user", "expected_password", "expected_known_hosts"),
    [
        ("Basic c2NydWZmeTp6ZXJvZw==", "scruffy", "zerog", ""),
        ("Basic", "", "", ""),
    ],
)
def test_parse_basic(
    response: str, expected_user: str, expected_password: str, expected_known_hosts: str
) -> None:
    user, password, known_hosts = authorize.parse_basic(response)
    assert user == expected_user
    assert password == expected_password
    assert known_hosts == expected_known_hosts


@pytest.mark.parametrize(
    "response",
    [
        "Basic!c2NydWZmeTp6ZXJvZw==",
        "Basic c2Nyd!!ZmeTp6ZXJvZw==",
        "Basic c2NydWZmeXplcm9n",
    ],
)
def test_parse_basic_invalid(response: str) -> None:
    with pytest.raises(ValueError):
        authorize.parse_basic(response)


def test_build_basic() -> None:
    assert authorize.build_basic("scruffy", "zerog") == "Basic c2NydWZmeTp6ZXJvZw=="


def test_build_basic_roundtrip() -> None:
    result = authorize.build_basic("user", "pass")
    user, password, known_hosts = authorize.parse_basic(result)
    assert user == "user"
    assert password == "pass"
    assert known_hosts == ""


def test_build_basic_none() -> None:
    result = authorize.build_basic(None, None)
    user, password, known_hosts = authorize.parse_basic(result)
    assert user == ""
    assert password == ""
    assert known_hosts == ""


def test_build_basic_with_known_hosts() -> None:
    result = authorize.build_basic("user", "pass", "host1.example.com ssh-rsa AAAA...")
    user, password, known_hosts = authorize.parse_basic(result)
    assert user == "user"
    assert password == "pass"
    assert known_hosts == "host1.example.com ssh-rsa AAAA..."


@pytest.mark.parametrize(
    ("response", "expected"),
    [
        ("Negotiate c2NydWZmeTp6ZXJvZw==", b"scruffy:zerog"),
        ("Negotiate", b""),
    ],
)
def test_parse_negotiate(response: str, expected: bytes) -> None:
    assert authorize.parse_negotiate(response) == expected


@pytest.mark.parametrize(
    "response",
    [
        "Negotiate!c2NydWZmeTp6ZXJvZw==",
        "Negotiate c2Nyd!!ZmeTp6ZXJvZw==",
    ],
)
def test_parse_negotiate_invalid(response: str) -> None:
    with pytest.raises(ValueError):
        authorize.parse_negotiate(response)


@pytest.mark.parametrize(
    ("data", "expected"),
    [
        (b"scruffy:zerog", "Negotiate c2NydWZmeTp6ZXJvZw=="),
        (None, "Negotiate"),
        (b"", "Negotiate"),
    ],
)
def test_build_negotiate(data: bytes | None, expected: str) -> None:
    assert authorize.build_negotiate(data) == expected


def test_build_negotiate_roundtrip() -> None:
    data = b"some binary \x00 data"
    result = authorize.build_negotiate(data)
    assert authorize.parse_negotiate(result) == data


@pytest.mark.parametrize(
    ("response", "expected_conversation", "expected_prompt"),
    [
        ("X-Conversation abcdefghi c2NydWZmeTp6ZXJvZw==", "abcdefghi", "scruffy:zerog"),
        ("X-Conversation abcdefghi", "abcdefghi", ""),
    ],
)
def test_parse_x_conversation(
    response: str, expected_conversation: str, expected_prompt: str
) -> None:
    conversation, prompt = authorize.parse_x_conversation(response)
    assert conversation == expected_conversation
    assert prompt == expected_prompt


def test_parse_x_conversation_invalid() -> None:
    with pytest.raises(ValueError):
        authorize.parse_x_conversation("X-Conversation abcdefghi c2NydW!!meTp6ZXJvZw==")


def test_build_x_conversation_with_prompt() -> None:
    result, conv = authorize.build_x_conversation("scruffy:zerog", "abcdefghi")
    assert result == "X-Conversation abcdefghi c2NydWZmeTp6ZXJvZw=="
    assert conv == "abcdefghi"


def test_build_x_conversation_no_prompt() -> None:
    result, conv = authorize.build_x_conversation("", "abcdefghi")
    assert result == "X-Conversation abcdefghi"
    assert conv == "abcdefghi"


def test_build_x_conversation_generate() -> None:
    result, conv = authorize.build_x_conversation("scruffy:zerog", None)
    assert conv
    assert "c2NydWZmeTp6ZXJvZw==" in result
    assert conv in result


def test_build_x_conversation_empty_invalid() -> None:
    with pytest.raises(ValueError):
        authorize.build_x_conversation("scruffy:zerog", "")


def test_build_x_conversation_roundtrip() -> None:
    prompt = "test prompt"
    result, conv = authorize.build_x_conversation(prompt, None)
    parsed_conv, parsed_prompt = authorize.parse_x_conversation(result)
    assert parsed_conv == conv
    assert parsed_prompt == prompt
